File: app/journal_environment.ml
type edge_insets =
  { left : float
  ; top : float
  ; right : float
  ; bottom : float
  }

type brightness =
  | Light
  | Dark

type orientation =
  | Portrait
  | Landscape

type snapshot =
  { viewport_width : float
  ; viewport_height : float
  ; device_pixel_ratio : float
  ; text_scale : float
  ; brightness : brightness
  ; platform : string
  ; locale : string
  ; safe_area : edge_insets
  ; keyboard_insets : edge_insets
  ; accessible_navigation : bool
  ; bold_text : bool
  ; invert_colors : bool
  ; disable_animations : bool
  ; reduced_motion : bool
  ; high_contrast : bool
  ; orientation : orientation
  ; pointer_kinds : int
  }

let equal_edge_insets left right =
  Float.equal left.left right.left
  && Float.equal left.top right.top
  && Float.equal left.right right.right
  && Float.equal left.bottom right.bottom
;;

let equal left right =
  Float.equal left.viewport_width right.viewport_width
  && Float.equal left.viewport_height right.viewport_height
  && Float.equal left.device_pixel_ratio right.device_pixel_ratio
  && Float.equal left.text_scale right.text_scale
  && left.brightness = right.brightness
  && String.equal left.platform right.platform
  && String.equal left.locale right.locale
  && equal_edge_insets left.safe_area right.safe_area
  && equal_edge_insets left.keyboard_insets right.keyboard_insets
  && Bool.equal left.accessible_navigation right.accessible_navigation
  && Bool.equal left.bold_text right.bold_text
  && Bool.equal left.invert_colors right.invert_colors
  && Bool.equal left.disable_animations right.disable_animations
  && Bool.equal left.reduced_motion right.reduced_motion
  && Bool.equal left.high_contrast right.high_contrast
  && left.orientation = right.orientation
  && Int.equal left.pointer_kinds right.pointer_kinds
;;

let fallback =
  { viewport_width = 0.
  ; viewport_height = 0.
  ; device_pixel_ratio = 1.
  ; text_scale = 1.
  ; brightness = Light
  ; platform = "unknown"
  ; locale = "en_US"
  ; safe_area = { left = 0.; top = 0.; right = 0.; bottom = 0. }
  ; keyboard_insets = { left = 0.; top = 0.; right = 0.; bottom = 0. }
  ; accessible_navigation = false
  ; bold_text = false
  ; invert_colors = false
  ; disable_animations = false
  ; reduced_motion = false
  ; high_contrast = false
  ; orientation = Portrait
  ; pointer_kinds = 0
  }
;;

let number (fields : (string * Yojson.Basic.t) list) name =
  match List.assoc_opt name fields with
  | Some (`Float value) -> Ok value
  | Some (`Int value) -> Ok (Float.of_int value)
  | _ -> Error ("environment " ^ name ^ " is missing or not a number")
;;

let boolean (fields : (string * Yojson.Basic.t) list) name =
  match List.assoc_opt name fields with
  | Some (`Bool value) -> Ok value
  | _ -> Error ("environment " ^ name ^ " is missing or not a bool")
;;

let string (fields : (string * Yojson.Basic.t) list) name =
  match List.assoc_opt name fields with
  | Some (`String value) -> Ok value
  | _ -> Error ("environment " ^ name ^ " is missing or not a string")
;;

let insets (fields : (string * Yojson.Basic.t) list) name =
  match List.assoc_opt name fields with
  | Some (`Assoc values) ->
    let ( let* ) = Result.bind in
    let* left = number values "left" in
    let* top = number values "top" in
    let* right = number values "right" in
    let* bottom = number values "bottom" in
    Ok { left; top; right; bottom }
  | _ -> Error ("environment " ^ name ^ " is missing or not an object")
;;

let decode_json (json : Yojson.Basic.t) =
  match json with
  | `Assoc fields ->
    let ( let* ) = Result.bind in
    let* viewport_width = number fields "viewportWidth" in
    let* viewport_height = number fields "viewportHeight" in
    let* device_pixel_ratio = number fields "devicePixelRatio" in
    let* text_scale = number fields "textScale" in
    let* platform = string fields "platform" in
    let* locale = string fields "locale" in
    let* safe_area = insets fields "safeArea" in
    let* keyboard_insets = insets fields "keyboardInsets" in
    let* accessible_navigation = boolean fields "accessibleNavigation" in
    let* bold_text = boolean fields "boldText" in
    let* invert_colors = boolean fields "invertColors" in
    let* disable_animations = boolean fields "disableAnimations" in
    let* reduced_motion = boolean fields "reducedMotion" in
    let* high_contrast = boolean fields "highContrast" in
    let* pointer_kinds = number fields "pointerKinds" in
    let* brightness =
      match List.assoc_opt "brightness" fields with
      | Some (`String "light") -> Ok Light
      | Some (`String "dark") -> Ok Dark
      | _ -> Error "environment brightness is missing or unsupported"
    in
    let* orientation =
      match List.assoc_opt "orientation" fields with
      | Some (`String "portrait") -> Ok Portrait
      | Some (`String "landscape") -> Ok Landscape
      | _ -> Error "environment orientation is missing or unsupported"
    in
    Ok
      { viewport_width
      ; viewport_height
      ; device_pixel_ratio
      ; text_scale
      ; brightness
      ; platform
      ; locale
      ; safe_area
      ; keyboard_insets
      ; accessible_navigation
      ; bold_text
      ; invert_colors
      ; disable_animations
      ; reduced_motion
      ; high_contrast
      ; orientation
      ; pointer_kinds = int_of_float pointer_kinds
      }
  | _ -> Error "environment snapshot must be a JSON object"
;;

let encode_json snapshot =
  let insets value =
    `Assoc
      [ "left", `Float value.left
      ; "top", `Float value.top
      ; "right", `Float value.right
      ; "bottom", `Float value.bottom
      ]
  in
  `Assoc
    [ "viewportWidth", `Float snapshot.viewport_width
    ; "viewportHeight", `Float snapshot.viewport_height
    ; "devicePixelRatio", `Float snapshot.device_pixel_ratio
    ; "textScale", `Float snapshot.text_scale
    ; ( "brightness"
      , `String (match snapshot.brightness with Light -> "light" | Dark -> "dark") )
    ; "platform", `String snapshot.platform
    ; "locale", `String snapshot.locale
    ; "safeArea", insets snapshot.safe_area
    ; "keyboardInsets", insets snapshot.keyboard_insets
    ; "accessibleNavigation", `Bool snapshot.accessible_navigation
    ; "boldText", `Bool snapshot.bold_text
    ; "invertColors", `Bool snapshot.invert_colors
    ; "disableAnimations", `Bool snapshot.disable_animations
    ; "reducedMotion", `Bool snapshot.reduced_motion
    ; "highContrast", `Bool snapshot.high_contrast
    ; ( "orientation"
      , `String
          (match snapshot.orientation with Portrait -> "portrait" | Landscape -> "landscape")
      )
    ; "pointerKinds", `Int snapshot.pointer_kinds
    ]
;;

File: swift/JournalEnvironment.swift
import Foundation
import SwiftUI
#if canImport(UIKit)
  import UIKit
#endif

/// Host environment snapshot pushed to OCaml as an LJP2 tag-24 event, field-
/// for-field what `Journal_environment.decode_json` expects (replaces the old
/// bonsai `environment_changed` protocol event).
struct JournalEnvironmentSample: Equatable {
  struct Insets: Equatable {
    var left = 0.0, top = 0.0, right = 0.0, bottom = 0.0
  }
  var viewportWidth = 0.0
  var viewportHeight = 0.0
  var devicePixelRatio = 1.0
  var textScale = 1.0
  var brightness = "light"
  var platform = "ios"
  var locale = "en"
  var safeArea = Insets()
  var keyboardInsets = Insets()
  var accessibleNavigation = false
  var boldText = false
  var invertColors = false
  var disableAnimations = false
  var reducedMotion = false
  var highContrast = false
  var orientation = "portrait"
  var pointerKinds = 0

  func jsonObject() -> [String: Any] {
    [
      "viewportWidth": viewportWidth,
      "viewportHeight": viewportHeight,
      "devicePixelRatio": devicePixelRatio,
      "textScale": textScale,
      "brightness": brightness,
      "platform": platform,
      "locale": locale,
      "safeArea": [
        "left": safeArea.left, "top": safeArea.top,
        "right": safeArea.right, "bottom": safeArea.bottom,
      ],
      "keyboardInsets": [
        "left": keyboardInsets.left, "top": keyboardInsets.top,
        "right": keyboardInsets.right, "bottom": keyboardInsets.bottom,
      ],
      "accessibleNavigation": accessibleNavigation,
      "boldText": boldText,
      "invertColors": invertColors,
      "disableAnimations": disableAnimations,
      "reducedMotion": reducedMotion,
      "highContrast": highContrast,
      "orientation": orientation,
      "pointerKinds": pointerKinds,
    ]
  }
}

/// Observes the application boundary and reports environment samples; the
/// platform deduplicates and pushes them through `journal_ocaml_platform_event`.
struct JournalEnvironmentObserver: View {
  let onSample: (JournalEnvironmentSample) -> Void
  @Environment(\.displayScale) private var displayScale
  @Environment(\.colorScheme) private var colorScheme
  @Environment(\.locale) private var locale
  @Environment(\.legibilityWeight) private var legibilityWeight
  @Environment(\.accessibilityVoiceOverEnabled) private var voiceOver
  @Environment(\.accessibilityInvertColors) private var invertColors
  @Environment(\.accessibilityReduceMotion) private var reduceMotion
  @Environment(\.colorSchemeContrast) private var contrast
  @Environment(\.layoutDirection) private var direction
  @ScaledMetric(relativeTo: .body) private var bodyScale: Double = 1

  private func sample(size: CGSize, safeArea: JournalEnvironmentSample.Insets)
    -> JournalEnvironmentSample
  {
    #if os(macOS)
      let platform = "macos"
      let pointers = 0x0e
    #else
      let platform = "ios"
      let pointers = 0x0f
    #endif
    return JournalEnvironmentSample(
      viewportWidth: size.width, viewportHeight: size.height,
      devicePixelRatio: displayScale, textScale: bodyScale,
      brightness: colorScheme == .dark ? "dark" : "light", platform: platform,
      locale: locale.identifier(.bcp47), safeArea: safeArea,
      keyboardInsets: .init(), accessibleNavigation: voiceOver,
      boldText: legibilityWeight == .bold, invertColors: invertColors,
      disableAnimations: reduceMotion, reducedMotion: reduceMotion,
      highContrast: contrast == .increased,
      orientation: size.width > size.height ? "landscape" : "portrait",
      // Host capability mask, not an inventory of connected input devices.
      pointerKinds: pointers)
  }

  var body: some View {
    Group {
      #if os(macOS)
        GeometryReader { geometry in
          let insets = geometry.safeAreaInsets
          let value = sample(
            size: geometry.size,
            safeArea: .init(
              left: direction == .leftToRight ? insets.leading : insets.trailing,
              top: insets.top,
              right: direction == .leftToRight ? insets.trailing : insets.leading,
              bottom: insets.bottom))
          Color.clear
            .onAppear { onSample(value) }
            .onChange(of: value) { _, value in onSample(value) }
        }
      #else
        UIKitEnvironmentProbe(preferences: sample(size: .zero, safeArea: .init())) {
          onSample($0)
        }
      #endif
    }
    .allowsHitTesting(false)
    .accessibilityHidden(true)
  }
}

#if canImport(UIKit)
  /// A zero-interaction UIView attached to the window that tracks viewport
  /// bounds, safe area, and keyboard occlusion (followsUndockedKeyboard),
  /// matching the bonsai UIKit environment probe.
  private final class JournalWindowGeometryView: UIView {
    struct Geometry {
      let size: CGSize
      let safeArea: JournalEnvironmentSample.Insets
      let keyboard: JournalEnvironmentSample.Insets
    }
    var onChange: ((Geometry) -> Void)?
    private var generation = UUID()
    private var scheduled = false
    private let tracker = UIView()

    override init(frame: CGRect) {
      super.init(frame: frame)
      isUserInteractionEnabled = false
      accessibilityElementsHidden = true
      backgroundColor = .clear
      keyboardLayoutGuide.followsUndockedKeyboard = true
      keyboardLayoutGuide.usesBottomSafeArea = false
      tracker.translatesAutoresizingMaskIntoConstraints = false
      addSubview(tracker)
      NSLayoutConstraint.activate([
        tracker.leadingAnchor.constraint(equalTo: keyboardLayoutGuide.leadingAnchor),
        tracker.trailingAnchor.constraint(equalTo: keyboardLayoutGuide.trailingAnchor),
        tracker.topAnchor.constraint(equalTo: keyboardLayoutGuide.topAnchor),
        tracker.bottomAnchor.constraint(equalTo: keyboardLayoutGuide.bottomAnchor),
      ])
    }
    @available(*, unavailable) required init?(coder: NSCoder) {
      fatalError("unavailable")
    }
    override func layoutSubviews() {
      super.layoutSubviews()
      schedule()
    }
    override func safeAreaInsetsDidChange() {
      super.safeAreaInsetsDidChange()
      schedule()
    }
    override func didMoveToWindow() {
      super.didMoveToWindow()
      schedule()
    }
    func invalidate() {
      generation = UUID()
      onChange = nil
    }
    func schedule() {
      guard !scheduled, window != nil, onChange != nil else { return }
      scheduled = true
      let generation = self.generation
      DispatchQueue.main.async { [weak self] in
        guard let self, self.generation == generation else { return }
        self.scheduled = false
        guard let window = self.window, self.window === window else { return }
        let keyboard = Self.keyboardInsets(
          viewport: self.bounds, occlusion: self.keyboardLayoutGuide.layoutFrame)
        let safe = window.safeAreaInsets
        self.onChange?(
          Geometry(
            size: window.bounds.size,
            safeArea: .init(
              left: safe.left, top: safe.top, right: safe.right, bottom: safe.bottom),
            keyboard: keyboard))
      }
    }

    /// Full-edge keyboard occlusion; partial-width keyboards do not count
    /// (same rule as bonsai's NativeKeyboardOcclusion).
    private static func keyboardInsets(viewport: CGRect, occlusion: CGRect)
      -> JournalEnvironmentSample.Insets
    {
      for rectangle in [viewport, occlusion] {
        guard
          [rectangle.origin.x, rectangle.origin.y, rectangle.size.width,
            rectangle.size.height]
            .allSatisfy(\.isFinite),
          rectangle.size.width >= 0, rectangle.size.height >= 0
        else { return .init() }
      }
      guard !viewport.isEmpty, !occlusion.isEmpty else { return .init() }
      let overlap = viewport.intersection(occlusion)
      guard !overlap.isNull, !overlap.isEmpty else { return .init() }
      if overlap.minX == viewport.minX, overlap.maxX == viewport.maxX {
        if overlap.maxY == viewport.maxY { return .init(bottom: overlap.height) }
        if overlap.minY == viewport.minY { return .init(top: overlap.height) }
      }
      if overlap.minY == viewport.minY, overlap.maxY == viewport.maxY {
        if overlap.minX == viewport.minX { return .init(left: overlap.width) }
        if overlap.maxX == viewport.maxX { return .init(right: overlap.width) }
      }
      return .init()
    }
  }

  private final class JournalEnvironmentAttachmentView: UIView {
    private var observer: JournalWindowGeometryView?
    var preferences = JournalEnvironmentSample()
    var onSample: ((JournalEnvironmentSample) -> Void)?

    override func didMoveToWindow() {
      super.didMoveToWindow()
      observer?.invalidate()
      observer = nil
      guard let window else { return }
      let observer = JournalWindowGeometryView(frame: window.bounds)
      self.observer = observer
      observer.onChange = { [weak self] geometry in
        guard let self else { return }
        var sample = self.preferences
        sample.viewportWidth = geometry.size.width
        sample.viewportHeight = geometry.size.height
        sample.orientation =
          geometry.size.width > geometry.size.height ? "landscape" : "portrait"
        sample.safeArea = geometry.safeArea
        sample.keyboardInsets = geometry.keyboard
        self.onSample?(sample)
      }
      window.insertSubview(observer, at: 0)
      observer.setNeedsLayout()
      observer.schedule()
    }
  }

  private struct UIKitEnvironmentProbe: UIViewRepresentable {
    let preferences: JournalEnvironmentSample
    let onSample: (JournalEnvironmentSample) -> Void

    func makeUIView(context: Context) -> JournalEnvironmentAttachmentView {
      let view = JournalEnvironmentAttachmentView()
      view.preferences = preferences
      view.onSample = onSample
      return view
    }
    func updateUIView(_ view: JournalEnvironmentAttachmentView, context: Context) {
      view.preferences = preferences
      view.onSample = onSample
    }
  }
#endif
